### PR TITLE
driver-openstack: Adds OpenStack driver

### PR DESCRIPTION
**Pull request by colin-campbell**

### CODE REVIEW

File: drivers/openstack/provider.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+	if p.userdata == nil {
+		p.userdata = userdata.T
+	}
+
+	if p.computeClient == nil {
+		opts, err := openstack.AuthOptionsFromEnv()
+
+		if err != nil {
+			panic(err)
```
Comment (tboerger) [inline drivers/openstack/provider.go]: Please don't panic, look at the Google provider, you can gracefully fail if you return an error

File: drivers/openstack/provider.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	if p.userdata == nil {
+		p.userdata = userdata.T
+	}
+
+	if p.computeClient == nil {
+		opts, err := openstack.AuthOptionsFromEnv()
```
Comment (tboerger) [inline drivers/openstack/provider.go]: You should use a different name here, opts is already an argument of this function.

File: drivers/openstack/provider.go (hunk 1 above)
Comment (colin-campbell) [reply drivers/openstack/provider.go]: Thanks. I was thinking about that, but while setupProvider() returns an error,  provider.New() does not and the calling function always returns nil.  Is it acceptable to change the signature of openstack.New() to return an error in this case?

Comment (colin-campbell) [reply drivers/openstack/provider.go]: Also. I notice that WithUserDataFile() option panics in all providers.

**Comment by colin-campbell** [comment]
Panics removed and some cleanup + tests added for provider and config.

**Comment by bradrydzewski** [comment]
This is awesome! The only feedback is that I would love to see improved test coverage on the create an destroy operations (following a pattern similar to [this](https://github.com/drone/autoscaler/blob/master/drivers/digitalocean/create_test.go))

**Comment by colin-campbell** [comment]
@bradrydzewski @tboerger I would love that too :) 
But I have spent the entire day attempting to get gock to handle the gophercloud results paging system in OS without any success. A single call to create a server can involve 10+ calls to the OS api listing images, secgroups, flavors, etc, and grok+gophercloud simply does not want to play ball. I shall continue, but it's not looking hopeful. 
I presume this is why the ec2 client lacks these tests too? :)

**Comment by bradrydzewski** [comment]
have you tested to make sure the code works with a live open stack instance?  If yes, we can merge and I can try to help add one or two gock tests in a separate patch. Perhaps if I put some example tests in place, you can help add a few more?

**Comment by JordanSussman** [comment]
This looks awesome!

Some initial feedback/thoughts on the creation step:

It is probably worthwhile to add the logic/option to utilize existing floating IPs instead of [creating](https://github.com/drone/autoscaler/pull/26/files#diff-12af315cedca5b09c056f960fe5c1fa3R30)/deleting them every time. The primary use case for this is related to having firewall rules on FIPs instead of whitelisting the entire OpenStack subnets.

Thoughts?

**Comment by tboerger** [comment]
But he's only creating a floating ip from a specific pool, isn't that enough?

**Comment by JordanSussman** [comment]
> But he's only creating a floating ip from a specific pool, isn't that enough?

Unfortunately, no creating/deleting the floating ip from an external pool isn't enough for our particular use case. To add further context/clarification around some OpenStack architecture:

- Floating IPs can be allocated from a particular pool to a project (tenant)
- Some of the external floating ip pools can be quite large (in our particular case we utilize some /21 and /22 networks)

Ideally, this code could handle the scenario where you give it a list of floating IPs already allocated to the project, instead of always getting random floating IPs from the external pool. Due to various regulatory reasons, we aren't able to blanket the entire external floating IP pool with firewall access. This requires us to allocate a select number of floating IPs from the pool, so we can assign firewall rules to them. 

On a semi-related note, you should at least consider switching from [openstack/compute/v2/extensions/floatingips](https://godoc.org/github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/floatingips) to [openstack/networking/v2/extensions/layer3/floatingips](https://godoc.org/github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/floatingips). The particular nova endpoints through `openstack/compute/v2/extensions/floatingips` have been deprecated for quite some time.

**Comment by colin-campbell** [comment]
Sure, I'm open to any ideas, but basically I want to get it merged and a full test suite done, before adding functionality. This is at the "scratching an itch"-stage at the moment. 
The OS API is considerably more tricky to deal with than some other providers, and I want a solid, functional base to work with.

**Comment by tboerger** [comment]
The first goal is to get it generally rolling on OpenStack of course. This should be already archived with some basic test coverage. When this is done we should be able to merge it and to add more flexibility for specific use cases like the one from @JordanSussman.

**Comment by tboerger** [comment]
Maybe you could create feature request issues for these improvements.

**Comment by JordanSussman** [comment]
Doesn't look like https://github.com/drone/autoscaler/issues is available. Should discussion move to https://discourse.drone.io/ instead?

**Comment by tboerger** [comment]
You are right, I forgot that the issues are disabled. Brad should decide where to put such issues.

**Comment by bradrydzewski** [comment]
I went ahead and merged the pull request, that way @JordanSussman can start working on a pull request. If you want to have any design discussion (around the feature) before you start coding, I recommend creating a placeholder pull request, and we can use the github comments to discuss further.

@colin-campbell if you have some time over the weekend (you mentioned previously) to try and get a gock test up and running that would be awesome. Try to pick the simplest function. Otherwise I can take a look early next week and see if I can get an example test in place.

**Comment by bradrydzewski** [comment]
I forgot to mention that the docs are in my private gogs repository. I will move these to GitHub so that we can get Open Stack (and Hetzner) properly documented.

**Comment by colin-campbell** [comment]
@bradrydzewski Tentatively, I think may have discovered  the gock issue. It seems gophercloud is extremely picky about the content-type which has to be explicitly set to application/json for the mocks. I'll open a new PR for the tests

**Pull request merged**